interpreter/cling/include/cling/Interpreter/Value.h

```diff
@@ -227,50 +192,51 @@ namespace cling {
-    template <typename T>
-    T getAs() const { return const_cast<Value*>(this)->getAs<T>(); }
-
-    template <typename T>
-    T*& getAs(T**) { return (T*&)getAs((void**)0); }
-    void*& getAs(void**) { return m_Storage.m_Ptr; }
-    double& getAs(double*) { return m_Storage.m_Double; }
-    long double& getAs(long double*) { return m_Storage.m_LongDouble; }
-    float& getAs(float*) { return m_Storage.m_Float; }
-    long long& getAs(long long*) { return m_Storage.m_LL; }
-    unsigned long long& getAs(unsigned long long*) { return m_Storage.m_ULL; }
-
-    void*& getPtr() { return m_Storage.m_Ptr; }
-    double& getDouble() { return m_Storage.m_Double; }
-    long double& getLongDouble() { return m_Storage.m_LongDouble; }
-    float& getFloat() { return m_Storage.m_Float; }
-    long long& getLL() { return m_Storage.m_LL; }
-    unsigned long long& getULL() { return m_Storage.m_ULL; }
-
+    // FIXME: If the cling::Value is destroyed and it handed out an address that
+    // might be accessing invalid memory.
+    void** getPtrAddress() { return &m_Storage.m_Ptr; }
     void* getPtr() const { return m_Storage.m_Ptr; }
-    double getDouble() const { return m_Storage.m_Double; }
-    long double getLongDouble() const { return m_Storage.m_LongDouble; }
-    float getFloat() const { return m_Storage.m_Float; }
-    long long getLL() const { return m_Storage.m_LL; }
-    unsigned long long getULL() const { return m_Storage.m_ULL; }
+    void setPtr(void* Val) { m_Storage.m_Ptr = Val; }
+
+#define _STRINGIFY(x) #x
+#define STRINGIFY(x) _STRINGIFY(x)
+#ifndef NDEBUG
+    // FIXME: Uncomment and debug the various type mismatches.
+    //#define ASSERT_TYPE_MISMATCH(name) AssertTypeMismatch(STRINGIFY(name))
+    #define ASSERT_TYPE_MISMATCH(name)
+#else
+    #define ASSERT_TYPE_MISMATCH(name)
+#endif // NDEBUG
+#define X(type, name)                                    \
+    type get##name() const {                             \
+      ASSERT_TYPE_MISMATCH(name);                        \
+      return m_Storage.m_##name;                         \
+    }                                                    \
+    void set##name(type Val) {                           \
+      ASSERT_TYPE_MISMATCH(name);                        \
+      m_Storage.m_##name = Val;                          \
+    }                                                    \
+
+  BUILTIN_TYPES
+
+#undef X
 
     /// \brief Get the value with cast.
     //
     /// Get the value cast to T. This is similar to reinterpret_cast<T>(value),
     /// casting the value of builtins (except void), enums and pointers.
     /// Values referencing an object are treated as pointers to the object.
     template <typename T>
-    T simplisticCastAs() const {
-      return CastFwd<T>::cast(*this);
+    T castAs() const {
+      if (needsManagedAllocation() || hasPointerType())
+        return (T) (uintptr_t) getAs<void*>();
+      if (hasBuiltinType())
+          return (T) getAs<T>();
```

Comment:
```suggestion
          return getAs<T>();
```
i.e. What is the function of the cast in the existing code?

Reply:
Yes, we can omit the `T` indeed.

Reply:
Then let's remove it :)